app/js/ZipManager.js

```diff
@@ -0,0 +1,98 @@
+const logger = require('logger-sharelatex')
+const UpdatesManager = require('./UpdatesManager')
+const DiffGenerator = require('./DiffGenerator')
+const DocumentUpdaterManager = require('./DocumentUpdaterManager')
+const PackManager = require('./PackManager')
+const yazl = require('yazl')
+const util = require('util')
+
+async function rewindDoc(projectId, docId, zipfile) {
+  logger.log({ projectId, docId }, 'rewinding document')
+
+  const [finalContent, version] =
+    await DocumentUpdaterManager.promises.getDocument(projectId, docId)
+
+  const id = docId.toString()
+
+  const contentEndPath = `${id}/content/end/${version}`
+  zipfile.addBuffer(Buffer.from(finalContent), contentEndPath)
+
+  const metadata = {
+    id,
+    version,
+    content: {
+      end: {
+        path: contentEndPath,
+        version,
+      },
+    },
+  }
+
+  // now rewind content
+  // TODO: retrieve updates incrementally
+  const updates = await PackManager.promises.getOpsByVersionRange(
+    projectId,
+    docId,
+    -1,
+    version
+  )
+
+  let content = finalContent
+  let v = version
+
+  metadata.updates = updates.map(update => {
+    const updatePath = `${id}/updates/${update.v}`
+
+    zipfile.addBuffer(Buffer.from(JSON.stringify(update)), updatePath, {
+      mtime: new Date(update.meta.start_ts),
+    })
+    try {
+      content = DiffGenerator.rewindUpdate(content, update)
```

Comment:
Would it be possible to use this `content` to add a `doc_length` property to each update, which is the length of the content on which the update operates (in characters)?

Reply:
Yes, I'll add that (using the code from [UpdateManager.js#L380-L413](https://github.com/overleaf/document-updater/blob/ecc7498720eca3f3b2a30c11d315c9491e49576b/app/js/UpdateManager.js#L380-L413) as a reference).